### PR TITLE
Greg

### PR DESCRIPTION
**Pull request by gpric024**
The test file is for testing the format of sentences. The load datasets python file demonstrates on how to get the wmt data (it does not actually download or do anything)

### CODE REVIEW

File: DataLoader.py [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+            file_type = path.split(".")[-1]
+            if file_type == "txt":
+                self.data = self.parse_txt(path)
+            elif file_type == "csv":
+                self.data = pd.read_csv(path, header=None)
+                #convert to list  of strings
+                self.data = self.data[0].tolist()
+            else:
+                raise Exception("Invalid file type")
+        elif data is not None:
```
Comment (Shaomei) [inline DataLoader.py:45]: When will this happen? Is it an expected use case to initialize the DataLoader with a list of sentences?

File: DataLoader.py [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+                #format each sentence in data
+                self.data = [self.fix_format(sentence) for sentence in data]
+            else:
+                raise Exception("Invalid data type, please pass in a list of sentences")
+        else:
+            raise Exception("Please pass in a path or data")
+
+    def parse_txt(self, path):
+        output = []
+        with open(path, "r") as f:
```
Comment (Shaomei) [inline DataLoader.py]:
```suggestion
        with open(path, "r", encoding="utf-8") as f:
```

Comment (Shaomei) [reply DataLoader.py]: It is often a good idea to specify the unicode encoding, unless you are parsing ascii/binary files.

File: DataLoader.py [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+        sentence = re.sub(r'‹|›', '"', sentence)
+        #if sentence begins and ends with quotes and there are only two, remove them
+        if sentence[0] == '"' and sentence[-1] == '"' and sentence.count('"') == 2:
+            sentence = sentence[1:-1]
+        elif sentence[0] == "'" and sentence[-1] == "'" and sentence.count("'") == 2:
+            sentence = sentence[1:-1]
+        return sentence
+
+    def save_as_txt(self, path):
+        with open(path, "w") as f:
```
Comment (Shaomei) [inline DataLoader.py]:
```suggestion
        with open(path, "w", encoding="utf-8") as f:
```

File: DataLoader.py [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+            for sentence in self.data:
+                f.write(f"{sentence}\n")
+        print(f"Saved {self.dataset_name} to {path}")
+        return
+    
+    def save_as_csv(self, path):
+        df = pd.DataFrame(self.data)
+        df.to_csv(path, index=False, header=False)
```
Comment (Shaomei) [inline DataLoader.py]:
```suggestion
        df.to_csv(path, index=False, header=False, encoding='utf-8')
```

Comment (Shaomei) [reply DataLoader.py]: Although utf-8 is the default encoding for python 3, it is not the case for python 2. Better to be explicit here.

File: DataLoader.py [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+        print(f"Saved {self.dataset_name} to {path}")
+        return
+    
+    def get_data(self):
+        return self.data
+
+    def create_deepcopy(self):
```
Comment (Shaomei) [inline DataLoader.py:98]: I guess this is the use case for the line 19, as I commented above. But it is unclear what the deepcopy is for? Is it for the cleaned up data? Some comments here can help clarify.

File: DyslexiaInjector.py [hunk 6]
```diff
@@ -0,0 +1,167 @@
+import random
+import DataLoader #Custom class
```
Comment (Shaomei) [inline DyslexiaInjector.py]: nit: I usually put the custom class below all the system classes.

File: DataLoader.py [hunk 7]
```diff
@@ -0,0 +1,76 @@
+import re
+import pandas as pd
+import copy
+class DataLoader:
```
Comment (Shaomei) [inline DataLoader.py:5]: Usually a good practice to use docstring here to explain the class and its usage. E.g. 

```
"""Loader for benchmarking datasets

(Longer explanation about the class)

Attributes:
    dataset_name:
    data:
"""
```

File: DyslexiaInjector.py [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+    
+    def injection_swap(self, p_start=0, p_end=1, step_size=0.1, save_path="", save_format="both"):
+        df_swap_results = pd.DataFrame(columns=["dataset","p1", "p2", "words_swapped", "letters_swapped", "words_changed", "sentences_changed"])
+        p1 = p_start
+        p2 = p_start
+        while p1 <= p_end:
+            #create deep copy of the data
+            if p1 == 0:
+                temp_load, results = self.injection_runner(self.load.create_deepcopy(), p1, p2)
+                df_swap_results.loc[len(df_swap_results)] = {"dataset":"wmt14_en", "p1":0, "p2":0, "words_swapped":results[0], "letters_swapped":results[1], "words_changed":results[2], "sentences_changed":results[3]}
```
Comment (Shaomei) [inline DyslexiaInjector.py]: nit: line too long, consider breaking into multiple lines.

File: DyslexiaInjector.py [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+                df_swap_results.loc[len(df_swap_results)] = {"dataset":"wmt14_en", "p1":p1, "p2":0, "words_swapped":results[0], "letters_swapped":results[1], "words_changed":results[2], "sentences_changed":results[3]}
+                self.saver(temp_load, save_path, p1, 0, save_format)
+
+                temp_load, results = self.injection_runner(self.load.create_deepcopy(), 0, p2)
+                df_swap_results.loc[len(df_swap_results)] = {"dataset":"wmt14_en", "p1":0, "p2":p2, "words_swapped":results[0], "letters_swapped":results[1], "words_changed":results[2], "sentences_changed":results[3]}
+                self.saver(temp_load, save_path, 0, p2, save_format)
+
+            #update the p1 and p2
+            p1 += step_size
+            p2 += step_size
```
Comment (Shaomei) [inline DyslexiaInjector.py]: Why are we incrementing p1 and p2 at the same time? Should this entire while loop be replaced with two nested for loops? Like:

```
for p1 in range(p_start, p_end, step_size):
    for p2 in range(p_start, p_end, step_size):
         temp_load, results = self.injection_runner(self.load.create_deepcopy(), p1, p2)
         df_swap_results.loc[len(df_swap_results)] = {"dataset":self.load.dataset_name,
                                                      "p1":p1,
                                                      "p2":p2,
                                                      "words_swapped":results[0],
                                                      "letters_swapped":results[1],
                                                      "words_changed":results[2],
                                                      "sentences_changed":results[3]}
         self.saver(temp_load, save_path, p1, p2, save_format)
```

File: DataLoader.py (hunk 1 above)
Comment (Shaomei) [reply DataLoader.py:45]: ah, I see, this is for the deepcopy..

File: DataLoader.py (hunk 5 above)
Comment (Shaomei) [reply DataLoader.py:98]: Okay, I see where this deepcopy is used now. Ignore this now. But perhaps there can be an cleaner way to do this.

File: DyslexiaInjector.py [hunk 10]
```diff
@@ -0,0 +1,167 @@
...
+    
+    def load_confusing_letters(self, path):
+        with open(path, "rb") as f:
+            confusing_letters_dict = pickle.load(f)
+            #close the file
+            f.close()
+        return confusing_letters_dict
+    
+    def injection_swap(self, p_start=0, p_end=1, step_size=0.1, save_path="", save_format="both"):
+        df_swap_results = pd.DataFrame(columns=["dataset","p1", "p2", "words_swapped", "letters_swapped", "words_changed", "sentences_changed"])
```
Comment (Shaomei) [inline DyslexiaInjector.py]: Consider changing p1, p2 to more meaningful names, such as p_letter, p_homophone.

File: DyslexiaInjector.py [hunk 11]
```diff
@@ -0,0 +1,167 @@
...
+        #first we need to split the sentence into a list of words
+        words = sentence.split()
+        #we need to keep track of the amount of words that were swapped
+        words_swapped = 0
+        #we also need to keep track of the amount of letters that were swapped
+        letters_swapped = 0
+        #we also need to keep track of the amount of words that were changed
+        words_changed = 0
+        for i in range(len(words)):
+            #if the word has any punctuation or capitalization, we skip it
```
Comment (Shaomei) [inline DyslexiaInjector.py]: We probably should not skip capitalized words or words with punctuation - even if it is harder to do the homophone confusion, we can still do the letter swapping, right?

File: DyslexiaInjector.py [hunk 12]
```diff
@@ -0,0 +1,167 @@
...
+                continue
+            word = words[i].lower().strip('".,?!:;()')
+            #create a copy of the word so we can check if it was changed
+            word_copy = word
+            #we need to check if the word is in the homophones dict
+            if word in self.homophones_dict.keys():
+                #we need to check if the word has any homophones
+                if len(self.homophones_dict[word]) > 0:
+                    #we need to check if we will swap the word with a homophone with probability p1
+
```
Comment (Shaomei) [inline DyslexiaInjector.py]: nit: remove empty line.

File: DyslexiaInjector.py [hunk 13]
```diff
@@ -0,0 +1,167 @@
...
+        for i in range(len(words)):
+            #if the word has any punctuation or capitalization, we skip it
+            if words[i].isalpha() == False:
+                continue
+            if words[i].islower() == False:
+                continue
+            word = words[i].lower().strip('".,?!:;()')
+            #create a copy of the word so we can check if it was changed
+            word_copy = word
+            #we need to check if the word is in the homophones dict
```
Comment (Shaomei) [inline DyslexiaInjector.py]: nit: remove "we need to" in the comments.

File: DyslexiaInjector.py [hunk 14]
```diff
@@ -0,0 +1,167 @@
...
+                    if random.random() < p1:
+                        print(p1)
+                        #we need to pick a random homophone from the list of homophones
+                        homophone = random.choice(self.homophones_dict[word])
+                        #we need to swap the word with the homophone
+                        words[i] = homophone
+                        #we need to update the amount of words that were swapped
+                        words_swapped += 1
+                        continue
+            p_letter = p2 #lower the probability of swapping a letter with a confusing letter each time we swap a letter
```
Comment (Shaomei) [inline DyslexiaInjector.py]: Move this comment to line 159

**Review by Shaomei — COMMENTED** [review]
Hi Greg,

I did a quick pass and shared some comments. Overall it looks great - much better than the notebook, I am very glad to see the code modularized and well-organized. Keep up the good work!

**Review by Shaomei — COMMENTED** [review]
Hi Greg,

I did a quick pass and shared some comments. Overall it looks great - much better than the notebook, I am very glad to see the code modularized and well-organized. Keep up the good work!

File: DataLoader.py (hunk 1 above)
Comment (gpric024) [reply DataLoader.py:45]: there is probably a more efficient way of doing this

File: DataLoader.py (hunk 5 above)
Comment (gpric024) [reply DataLoader.py:98]: There is definitely a cleaner way of doing it. I will see what I can think of

File: DyslexiaInjector.py (hunk 6 above)
Comment (gpric024) [reply DyslexiaInjector.py]: Got it. Will try to remember

File: DyslexiaInjector.py (hunk 10 above)
Comment (gpric024) [reply DyslexiaInjector.py]: Done

File: DyslexiaInjector.py (hunk 8 above)
Comment (gpric024) [reply DyslexiaInjector.py]: Done

**Comment by gpric024** [comment]
Fixes will be present in next pull request

**Pull request closed without merging**